common/htlc_wire.c

```diff
@@ -84,9 +87,12 @@ struct failed_htlc *fromwire_failed_htlc(const tal_t *ctx, const u8 **cursor, si
 	struct failed_htlc *failed = tal(ctx, struct failed_htlc);
 
 	failed->id = fromwire_u64(cursor, max);
-	failed->malformed = fromwire_u16(cursor, max);
+	failed->failcode = fromwire_u16(cursor, max);
 	failreason_len = fromwire_u16(cursor, max);
-	failed->failreason = tal_arr(failed, u8, failreason_len);
+	if (failreason_len)
+		failed->failreason = tal_arr(failed, u8, failreason_len);
+	else
+		failed->failreason = NULL;
```

Comment:
Isn't `fromwire_u8_array` technically allowed to dereference `failed->failreason`, even though the length is 0 in this case? Can we more the call to `fromwire_u8_array` into the `if`-branch?

Reply:
Not quite: the argument would be that the compiler could assume the latter branch never happens because you can't hand "NULL" to memcpy (as done in fromwire).

But this fights with Kernighan and Plauger's "Make sure special cases are truly special".  We could, at best, hide it inside fromwire() and expicitly test for 0-length there, but TBH compilers don't currently actually assume the NULL-to-memcpy-is-impossible because, well, I am right :)